assignments/tasks.go

```diff
@@ -0,0 +1,284 @@
+	tasksToBeCreated := []*pb.Task{}
+	tasksToBeUpdated := []*pb.Task{}
+	taskMap := make(map[string]*pb.Task)
+	for _, task := range tasks {
+		if task.AssignmentOrder == assignment.Order {
+			taskMap[task.Name] = task
+		}
+	}
+
+	existingTasks, err := db.GetTasks(&pb.Task{AssignmentID: assignment.GetID()})
+	if err != nil {
+		return err
+	}
+
+	for _, existingTask := range existingTasks {
+		task, ok := taskMap[existingTask.Name]
+		if !ok {
+			// There exists a task in db, that is not represented by a task found in scm.
+			db.DeleteTask(existingTask)
+			continue
+		}
+		if !(task.Title == existingTask.Title && task.Body == existingTask.Body) {
+			// Task has been changed
+			existingTask.Title = task.Title
+			existingTask.Body = task.Body
+			tasksToBeUpdated = append(tasksToBeUpdated, existingTask)
+		}
+		delete(taskMap, existingTask.Name)
+	}
+
+	// Only tasks that there is no existing record of remains
+	for _, task := range taskMap {
+		tasksToBeCreated = append(tasksToBeCreated, task)
+	}
+
+	err = db.CreateTasks(tasksToBeCreated)
+	if err != nil {
+		return err
+	}
+	err = db.UpdateTasks(tasksToBeUpdated)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// synchronizeIssues synchronizes database and scm with issues based on tasks found
+func synchronizeIssues(c context.Context, db database.Database, course *pb.Course, s scm.SCM, repo *pb.Repository, tasks []*pb.Task) error {
+	issuesToBeCreated := []*pb.Issue{}
+	issuesToBeUpdated := []*pb.Issue{}
+	tasksMap := make(map[string]*pb.Task)
+	for _, task := range tasks {
+		tasksMap[task.Name] = task
+	}
+
+	// Loops through existing issues on repo.
+	for _, issue := range repo.Issues {
+		task, ok := tasksMap[issue.Name]
+		if !ok {
+			// What should happen if task does not exist for issue?
+			continue
+		}
+		// Might be necessary to check if the issue is closed here, otherwise we might end up editing already closed issues.
+		// If this becomes a necessity, then the issue data-record must have a closed (true/false) field, or a status field.
+		if !(task.Title == issue.Title && task.Body == issue.Body) {
+			// Issue needs to be updated here
+			issue.Title = task.Title
+			issue.Body = task.Body
+			issuesToBeUpdated = append(issuesToBeUpdated, issue)
+		}
+		delete(tasksMap, issue.Name)
+	}
+
+	// Only tasks that do not have an issue with them remain.
+	for _, task := range tasksMap {
+		// Creates the actual issue on a scm
+		scmIssue, err := createScmIssue(c, s, course, repo, task)
+		if err != nil {
+			return err
+		}
+		// Creates issue to be saved in db
+		issue, err := createIssue(c, repo, task, scmIssue)
+		if err != nil {
+			return err
+		}
+		issuesToBeCreated = append(issuesToBeCreated, issue)
+	}
+
+	// Updates issues on scm.
+	for _, issue := range issuesToBeUpdated {
+		_, err := updateScmIssue(c, s, course, repo, issue)
+		if err != nil {
+			return err
+		}
+	}
+
+	err := db.CreateIssues(issuesToBeCreated)
```

Comment:
The `CreateIssue` and `UpdateIssue` database methods are only used once; meaning it might be better to just combine them somehow into a single method to avoid bloating the database interface. Recall also error handling and returning error as explained for the other sync func above.